### PR TITLE
Plfm 5960

### PR DESCRIPTION
**Pull request by nlgilber**

### CODE REVIEW

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/NodeDAOImpl.java [hunk 1]
```diff
@@ -353,17 +395,16 @@
...
 	private static final int NODE_VERSION_LIMIT_BY_FILE_MD5 = 200;
 	private static final String SELECT_NODE_VERSION_BY_FILE_MD5 =
 			ENTITY_HEADER_SELECT
-			+ " FROM " + TABLE_REVISION + " R, " + TABLE_FILES + " F, "+TABLE_NODE+" N"
-			+ " WHERE R."+COL_REVISION_OWNER_NODE+" = N."+COL_NODE_ID+" AND  R." + COL_REVISION_FILE_HANDLE_ID + " = F." + COL_FILES_ID
+			+ " FROM " + TABLE_REVISION + " R, " + TABLE_FILES + " F, " + TABLE_NODE + " N"
+			+ " WHERE R." + COL_REVISION_OWNER_NODE + " = N."+COL_NODE_ID+" AND  R." + COL_REVISION_FILE_HANDLE_ID + " = F." + COL_FILES_ID
 			+ " AND F." + COL_FILES_CONTENT_MD5 + " = :" + COL_FILES_CONTENT_MD5
-			+ " LIMIT " + (NODE_VERSION_LIMIT_BY_FILE_MD5 + 1);
+			+ " ORDER BY N." + COL_NODE_CREATED_ON + " DESC"
```
Comment (marcomarasca) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/NodeDAOImpl.java]: This order by will effectively switch (or at least different) the order compared to the current implementation (which we can say it's undefined, but most likely ordered by the node id). An order by should still be specified when applying a limit, but to keep it "backward compatible" I'd probably order by the N.ID.

File: services/repository/src/main/java/org/sagebionetworks/repo/web/controller/EntityController.java [hunk 2]
```diff
@@ -1424,7 +1423,13 @@ public void filePreviewRedirectURLForVersion(
 
 	/**
 	 * Gets all FileEntities whose file's MD5 is the same as the specified MD5
-	 * string.
+	 * string. NOTE: This will return a subset of at most 200 entity headers that is the
+	 * intersection of the set of the most recently created entity headers with the specified MD5 string and
```
Comment (marcomarasca) [inline services/repository/src/main/java/org/sagebionetworks/repo/web/controller/EntityController.java]: We can simply say that it returns at most 200 entities matching the given md5 for which the user has read access.

File: services/repository/src/main/java/org/sagebionetworks/repo/web/controller/EntityController.java [hunk 3]
```diff
@@ -1424,7 +1423,13 @@ public void filePreviewRedirectURLForVersion(
...
+	 * the set of entities which the user has access to. To look at MD5 values of entities
+	 * in a specific project, create a file view.
```
Comment (marcomarasca) [inline services/repository/src/main/java/org/sagebionetworks/repo/web/controller/EntityController.java]: Not necessarily for a project only, I would just mention that for more flexibility a file view that includes the MD5 can be used.

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/NodeDAOImplTest.java [hunk 4]
```diff
@@ -2339,80 +2340,35 @@ public void testCreateTableNode() throws DatastoreException, InvalidModelExcepti
 		assertEquals(expected, n1.getColumnModelIds());
 	}
 
+	/**
+	 * PLFM-5960
+	 * @throws Exception
+	 */
 	@Test
 	public void testGetEntityHeaderByMd5() throws Exception {
```
Comment (marcomarasca) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/NodeDAOImplTest.java]: We can rename this to testGetEntityHeaderByMd5WithNoMatch

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/NodeDAOImplTest.java [hunk 5]
```diff
@@ -2339,80 +2340,35 @@ public void testCreateTableNode() throws DatastoreException, InvalidModelExcepti
...
-
-		// Add a new node with no file handle
-		Node node2 = NodeTestUtils.createNew("testGetEntityHeaderByMd5 node 2", creatorUserGroupId);
-		final String node2Label1 = "Node 2 version label 1";
-		node1.setVersionLabel(node2Label1);
-		final String id2 = nodeDao.createNew(node2);
-		assertNotNull(id2);
-		toDelete.add(id2);
-		node2.setId(id2);
+		for(int i = 0; i < 202; i++) {
```
Comment (marcomarasca) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/NodeDAOImplTest.java]: Let us separate this into its own test, e.g. testGetEntityHeaderByMd5WithOver200Macthing

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/NodeDAOImplTest.java [hunk 6]
```diff
@@ -2339,80 +2340,35 @@ public void testCreateTableNode() throws DatastoreException, InvalidModelExcepti
...
+			// Add a node with a file handle
+			Node curr = NodeTestUtils.createNew("testGetEntityHeaderByMd5 node " + i, creatorUserGroupId);
+			curr.setFileHandleId(fileHandle3.getId());
+			final String nodeLabel = "" + i;
+			curr.setVersionLabel(nodeLabel);
+			final String id = nodeDao.createNew(curr);
+			curr = nodeDao.getNode(id);
+			assertNotNull(id);
+			toDelete.add(id);
+			curr.setId(id);
```
Comment (marcomarasca) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/NodeDAOImplTest.java]: This is not needed.

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/NodeDAOImplTest.java [hunk 7]
```diff
@@ -2339,80 +2340,35 @@ public void testCreateTableNode() throws DatastoreException, InvalidModelExcepti
...
-		toDelete.add(id2);
-		node2.setId(id2);
+		for(int i = 0; i < 202; i++) {
+			// Add a node with a file handle
+			Node curr = NodeTestUtils.createNew("testGetEntityHeaderByMd5 node " + i, creatorUserGroupId);
+			curr.setFileHandleId(fileHandle3.getId());
+			final String nodeLabel = "" + i;
+			curr.setVersionLabel(nodeLabel);
+			final String id = nodeDao.createNew(curr);
+			curr = nodeDao.getNode(id);
```
Comment (marcomarasca) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/NodeDAOImplTest.java]: This is not used.

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/NodeDAOImplTest.java [hunk 8]
```diff
@@ -2339,80 +2340,35 @@ public void testCreateTableNode() throws DatastoreException, InvalidModelExcepti
...
+			assertNotNull(id);
+			toDelete.add(id);
+			curr.setId(id);
+		}
 
-		results = nodeDao.getEntityHeaderByMd5(fileHandle.getContentMd5());
+		results = nodeDao.getEntityHeaderByMd5(fileHandle3.getContentMd5());
```
Comment (marcomarasca) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/NodeDAOImplTest.java]: Please add // Call under test

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/NodeDAOImplTest.java [hunk 9]
```diff
@@ -2339,80 +2340,35 @@ public void testCreateTableNode() throws DatastoreException, InvalidModelExcepti
...
 		assertNotNull(results);
-		assertEquals(2, results.size());
-		assertEquals(id1, results.get(0).getId());
-		assertEquals(id1, results.get(1).getId());
+		assertEquals(200, results.size());
```
Comment (marcomarasca) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/NodeDAOImplTest.java]: We can reuse the constant

**Review by marcomarasca — CHANGES_REQUESTED** [review]

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/NodeDAOImplTest.java (hunk 6 above)
Comment (nlgilber) [reply lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/NodeDAOImplTest.java]: must add the id to the toDelete list which deletes values in the after() section of the test

File: lib/models/src/main/java/org/sagebionetworks/repo/model/NodeDAO.java [hunk 10]
```diff
@@ -31,7 +31,8 @@
 	 * the value to pass into the node to remove the generatedBy link between node and activity
 	 * 
 	 */
-	public static String DELETE_ACTIVITY_VALUE = "-1";
+	String DELETE_ACTIVITY_VALUE = "-1";
+	int NODE_VERSION_LIMIT_BY_FILE_MD5 = 200;
```
Comment (marcomarasca) [inline lib/models/src/main/java/org/sagebionetworks/repo/model/NodeDAO.java:35]: You might want to add some documentation

**Review by marcomarasca — APPROVED** [review]

**Pull request merged**